### PR TITLE
change to node20 for all actions

### PR DESCRIPTION
**Pull request by nickfyson**
This PR upgrades all the actions to use `node20`, and bumps the major version to `v3`.

It also adds a new PR check to confirm the consistency of node versions across all actions, and ensure that we do not change this version in backport PRs to older releases.

### Merge / deployment checklist

- [x] Confirm this change is backwards compatible with existing workflows.
- [x] Confirm the [readme](https://github.com/github/codeql-action/blob/main/README.md) has been updated if necessary.
- [x] Confirm the [changelog](https://github.com/github/codeql-action/blob/main/CHANGELOG.md) has been updated if necessary.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -4,7 +4,7 @@ See the [releases page](https://github.com/github/codeql-action/releases) for th
 
 ## [UNRELEASED]
 
-No user facing changes.
+- [v3+ only] The CodeQL Action now runs on Node.js v20. [#1000](https://github.com/github/codeql-action/pull/2006)
```
Comment (aeisenberg) [inline CHANGELOG.md:7]:
```suggestion
- [v3+ only] The CodeQL Action now runs on Node.js v20. [#2006](https://github.com/github/codeql-action/pull/2006)
```

File: foobar.txt
Comment (aeisenberg) [inline foobar.txt:1]: Accidental commit?

**Review by aeisenberg — COMMENTED** [review]
I think the changes you made are good for bumping the node version of the actions. However, once we merge this, we can no longer release `v2` of the action until we also update the release scripts to what we had before when we released `v1` and `v2` (now will be `v2` and `v3`). That's a dicey state to be in.

So, either we can recreate the release scripts and merge it first, or we do them at the same time. I can't remember how we did it last time.

**Comment by aeisenberg** [comment]
Right...it was this PR that added the v2->v1->main release steps https://github.com/github/codeql-action/pull/995. It was merged before the node version changes were merged in #1000.

File: foobar.txt
Comment (nickfyson) [reply foobar.txt:1]: :thanks:

**Comment by aeisenberg** [comment]
This PR https://github.com/github/codeql-action/pull/1482 is the revert PR that removed the v1 release process.

**Review by henrymercer — COMMENTED** [review]
Looks good.  Let's wait for the CodeQL 2.15.4 release to complete before merging this to reduce the risk of that release.

One potential safety check we could add is a PR check that fails if the release branch is `v2` and the Node version of any Actions is Node 20.  This could help make sure  (a) we don't forget to switch the v2 branch back to running on `node16` the first time we release it, and (b) any new Action we add to `main` runs on `node16` in the `v2` branch.

File: .github/workflows/pr-checks.yml [hunk 2]
```diff
@@ -15,13 +15,39 @@ jobs:
     runs-on: ubuntu-latest
     timeout-minutes: 45
 
+    strategy:
+      matrix:
+        node-types-version: [16.11, current] # run tests on 16.11 while codeql-action v2 is still supported
```
Comment (henrymercer) [inline .github/workflows/pr-checks.yml:20]:
```suggestion
        node-types-version: [16.11, current] # run tests on 16.11 while CodeQL Action v2 is still supported
```

File: .github/actions/check-sarif/action.yml [hunk 3]
```diff
@@ -16,5 +16,5 @@ inputs:
       Comma separated list of query ids that should NOT be included in this SARIF file.
 
 runs:
-  using: node16
+  using: 'node20'
```
Comment (henrymercer) [inline .github/actions/check-sarif/action.yml:19]: Nit: I think these quotes are redundant throughout, but maybe now's not the time to remove them.

**Review by henrymercer — APPROVED** [review]

**Review by henrymercer — DISMISSED** [review]

File: .github/workflows/pr-checks.yml [hunk 4]
```diff
@@ -114,3 +114,44 @@ jobs:
           # we won't be able to find them on Windows.
           npm config set script-shell bash
           npm test
+
+  check-node-version:
+    if: ${{ github.event.pull_request }}
+    name: Check node version consistency
```
Comment (henrymercer) [inline .github/workflows/pr-checks.yml:120]:
```suggestion
    name: Check Action Node versions
```

File: .github/workflows/pr-checks.yml [hunk 5]
```diff
@@ -114,3 +114,44 @@ jobs:
...
+    runs-on: ubuntu-latest
+    timeout-minutes: 45
+    env:
+      BASE_REF: ${{ github.base_ref }}
+
+    steps:
+      - uses: actions/checkout@v4
+      - id: head-version
+        name: check HEAD node version
```
Comment (henrymercer) [inline .github/workflows/pr-checks.yml:129]:
```suggestion
        name: Verify all Actions use the same Node version
```

File: .github/workflows/pr-checks.yml [hunk 6]
```diff
@@ -114,3 +114,44 @@ jobs:
...
+        run: |
+          NODE_VERSION=$(find . -name "action.yml" -exec yq -e '.runs.using' {} \; | grep node | sort | uniq)
+          echo "NODE_VERSION: ${NODE_VERSION}"
+          if [[ $(echo "$NODE_VERSION" | wc -l) -gt 1 ]]; then
+            echo "Error: More than one node version used in actions."
```
Comment (henrymercer) [inline .github/workflows/pr-checks.yml:134]:
```suggestion
            echo "::error::More than one node version used in 'action.yml' files."
```

File: .github/workflows/pr-checks.yml [hunk 7]
```diff
@@ -114,3 +114,44 @@ jobs:
...
+            exit 1
+          fi
+          echo "node_version=${NODE_VERSION}" >> $GITHUB_OUTPUT
+
+      - id: checkout-base
+        name: check out base ref for backport check
```
Comment (henrymercer) [inline .github/workflows/pr-checks.yml:140]:
```suggestion
        name: 'Backport: Check out base ref'
```

File: .github/workflows/pr-checks.yml [hunk 8]
```diff
@@ -114,3 +114,44 @@ jobs:
...
+        if: ${{ startsWith(github.head_ref, 'backport-') }}
+        uses: actions/checkout@v4
+        with:
+          ref: ${{ env.BASE_REF }}
+
+      - name: compare with node version on base ref for backport check
```
Comment (henrymercer) [inline .github/workflows/pr-checks.yml:146]:
```suggestion
      - name: 'Backport: Verify Node versions unchanged'
```

File: .github/workflows/pr-checks.yml [hunk 9]
```diff
@@ -114,3 +114,44 @@ jobs:
...
+        if: steps.checkout-base.outcome == 'success'
+        env:
+          HEAD_VERSION: ${{ steps.head-version.outputs.node_version }}
+        run: |
+          BASE_VERSION=$(find . -name "action.yml" -exec yq -e '.runs.using' {} \; | grep node | sort | uniq)
+          echo "HEAD_VERSION: ${HEAD_VERSION}"
+          echo "BASE_VERSION: ${BASE_VERSION}"
+          if [[ "$BASE_VERSION" != "$HEAD_VERSION" ]]; then
+            echo "Error: Cannot change node version in a backport PR."
```
Comment (henrymercer) [inline .github/workflows/pr-checks.yml:155]:
```suggestion
            echo "::error::Cannot change the Node version of an Action in a backport PR."
```

**Review by henrymercer — COMMENTED** [review]
Minor suggestions to the new PR check, looks good overall :)

**Comment by nickfyson** [comment]
NB – The introduction of a matrix for the Check JS job means that `Check JS` has been replaced by `Check JS (16.11)` and `Check JS (current)`.

Once this PR is approved and ready to merge I will remove `Check JS` as a required check, and once merged I will run the `update-required-checks.sh` script in order to make the new checks required.

**Pull request merged**